### PR TITLE
Remove `BlockReader::read_blk` in favour of `BlockCursor`

### PR DESCRIPTION
**Pull request by arpad-m**
## Problem

We want to make `read_blk` an async function, but outside of `async_trait`, which allocates, and nightly features, we can't use async fn's in traits.

## Summary of changes

* Remove all uses of `BlockReader::read_blk` in favour of using block cursors.
* Introduce a `BlockReaderRef` enum that lists all implementors of `BlockReader::read_blk`.
* Remove `BlockReader::read_blk` and move its implementations into inherent functions on the types instead.

We don't turn `read_blk` into an async fn yet, for that we also need to modify the page cache. So this is a preparatory PR, albeit an important one.

Part of #4743.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by arpad-m** [comment]
Rebased after merge of #4994

File: pageserver/src/tenant/block_io.rs
Comment (koivunej) [inline pageserver/src/tenant/block_io.rs:75]: Could you restore this?

File: pageserver/src/tenant/storage_layer/delta_layer.rs
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:1007]: Why?

File: pageserver/src/tenant/storage_layer/delta_layer.rs
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:1031]: This bound should be possible to get rid of. Though, I did try that unsure why it wasn't possible.

File: pageserver/src/tenant/storage_layer/delta_layer.rs
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/delta_layer.rs:1037]: Because this is ok, then it must be ok to keep the old versions, but it need to add the `'a` to ValueRef like it is right now, but also add the `T` from `Adapter<T>` which needs to be on `BlockCursor<'a, T>`.

File: pageserver/src/tenant/block_io.rs
Comment (arpad-m) [reply pageserver/src/tenant/block_io.rs:75]: I would argue that for #4938, one should add another variant to this enum, instead of making the entire enum generic. Making the enum generic would mean that we have to make the return type an associated type again. I'm not a fan of that, as it makes things more complicated than they should be.

Ideally we'd wait for #4938 to be merged, and then I can implement my proposed approach on top of that. That way would be the least disruptive for you.

Comment (arpad-m) [reply pageserver/src/tenant/block_io.rs:75]: I mean, if `BlockCursor` became generic over the type, the problem is that we'd have to make the type be an associated type of the trait, and that would seem extremely ugly to me, as we don't have associated type defaults yet, so it would have to be repeated on all impls of the `BlockReader` trait.

Comment (koivunej) [reply pageserver/src/tenant/block_io.rs:75]: Not needed anymore.

File: pageserver/src/tenant/storage_layer/delta_layer.rs
Comment (koivunej) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1007]: Not needed anymore.

File: pageserver/src/tenant/storage_layer/delta_layer.rs
Comment (koivunej) [reply pageserver/src/tenant/storage_layer/delta_layer.rs:1034]: Not needed anymore.

File: pageserver/src/tenant/block_io.rs
Comment (koivunej) [inline pageserver/src/tenant/block_io.rs:71]: Could add a quick note on what is the purpose of this. If we read this many months from now it might look a bit extra.

```suggestion
/// Provides the ability to read blocks from different sources, alternative
/// to using traits which are problematic for asyncification.
pub(crate) enum BlockReaderRef<'a> {
```

Or something like this.

File: pageserver/src/tenant/block_io.rs
Comment (koivunej) [inline pageserver/src/tenant/block_io.rs:81]: Do you know if this makes any difference? I could imagine it is helpful with these kinds of structures where there's a lot of fan out but I'd imagine llvm seeing through to the only choice, but is it unable to do that without this hint?

**Review by koivunej — APPROVED** [review]
Adapter, DeltaLayerInner::load_keys now have extra code, but this should be ok to leave, I'll clean them up.

File: pageserver/src/tenant/block_io.rs
Comment (arpad-m) [reply pageserver/src/tenant/block_io.rs:81]: Yeah as we don't have any `dyn BlockReader` cases on master, all of the cases where `BlockReaderRef` is used is in cases where we know the concrete type that implements `BlockReader`, and thus we also know the implementation of the `block_cursor` function, which always creates a specific `BlockReaderRef`. In other words, `BlockReaderRef::read_blk` will always only resolve to one specific enum variant at each call site.

Ideally LLVM realizes this as well and spares the dispatch overhead in `read_blk` (which is non-trivial due to the (upcoming) async state machines). For this to happen, LLVM needs insight into the functions involved. Many LLVM optimization passes are implemented only to apply to the same function, and if LLVM doesn't inline something, that pass won't recognize the case, so that's why I put the inlining hint here.

**Pull request merged**
